### DISCUSSION TITLE
Best reconstruction strategy for a 3D volume, either a for loop over 2D or straight 3D?

### DISCUSSION THREAD

**Discussion by aljosahafner** [astra-toolbox/astra-toolbox#471] [Q&A] [answered]
Hello,

I inherited a legacy CT project where the input file is a 3D volume of shape (# sinograms, # projections, # det. size). Currently we do the final reconstruction using the following (reduced, simplified) pseudo-algorithm:

```
from joblib import Parallel, delayed

Parallel(n_jobs=numCPUs, verbose=1)(delayed(reconstruct_joblib)(i, dset, params) for i in range(start_slice, end_slice+1))
```

This way, the function `reconstruct_joblib` gets a sinogram `i` and then reconstructs it with a 2D algorithm (`dset` is a memmap array):

```
def reconstruct_joblib(params)
    im = reconstruct_astra(dset[i,:,:], params)
```

```
from numpy import float32, linspace
import astra

def reconstruct(im, angles, method, filter_type, cor_shift):
    vol_geom = astra.create_vol_geom(im.shape[1], im.shape[1])
    proj_geom = astra.create_proj_geom('parallel', 1.0, im.shape[1], linspace(0, angles, im.shape[0], False))

    # Projection geometry with shifted center of rotation:
    proj_geom = astra.geom_postalignment(proj_geom, cor_shift);
    
    # Create a data object for the reconstruction
    rec_id = astra.data2d.create('-vol', vol_geom)  

    # We now re-create the sinogram data object:
    sinogram_id = astra.data2d.create('-sino', proj_geom, im)

    # Create configuration:
    cfg = astra.astra_dict(method)
    cfg['ReconstructionDataId'] = rec_id
    cfg['ProjectionDataId'] = sinogram_id
    cfg['FilterType'] = filter_type 

    # Create and run the algorithm object from the configuration structure
    alg_id = astra.algorithm.create(cfg)
    astra.algorithm.run(alg_id, 1)

    # Get the result
    rec = astra.data2d.get(rec_id)

    # Clean up:
    astra.algorithm.delete(alg_id)
    astra.data2d.delete(rec_id)
    astra.data2d.delete(sinogram_id)
    
    return rec
```

Would it be better to change the code to a 3D volume (some hint how best to do it is appreciated) or keep it this way? I am reconstructing with CUDA of course.

**Comment by askorikov** [comment]
Hi! We unfortunately don't have a full 3D implementation of FBP algorithm. In any case, it's not very likely that you will get any significant performance increase from the 3D version. That would be the case for small reconstruction dimensions (smaller than about 512x512 pixels, depending on the hardware), when 2D algorithm for each slice severely underutilizes the GPU. On the other hand, the advantage of slice-by-slice reconstruction is that you basically don't have to worry about running out of GPU memory.

**Comment by aljosahafner** [comment]
Thanks for the answer. So you suggest to stay in the current implementation? Perhaps a further improvement would be to work simultaneously on multiple GPUs for different slices. What do you think?

If there are e. g. 3600 images of 2k x 2k, then it takes 10-20 minutes to complete the reconstruction, so I am searching for a faster algorithm to process such a volume...

**Comment by askorikov** [comment, marked as answer]
Yes indeed. You can split you reconstruction into subsets by slices and do them on separate GPUs with multithreading and [`set_gpu_index`](https://astra-toolbox.com/apiref/astra.html#astra.astra.set_gpu_index). Something like:

```python
import astra
import numpy as np
from multiprocessing.pool import ThreadPool


def fbp_subset(gpu_id, subset):
    astra.astra.set_gpu_index(gpu_id)
    proj_geom = astra.create_proj_geom('parallel', 1.0, N, angles)
    vol_geom = astra.create_vol_geom(N, N)
    rec_id = astra.data2d.create('-vol', vol_geom)
    cfg = astra.astra_dict('FBP_CUDA')
    cfg['ReconstructionDataId'] = rec_id
    rec = np.empty([subset.shape[0], N, N], dtype=np.float32)
    for i in range(subset.shape[0]):
        sinogram_id = astra.data2d.create('-sino', proj_geom, subset[i, ...])
        cfg['ProjectionDataId'] = sinogram_id
        alg_id = astra.algorithm.create(cfg)
        astra.algorithm.run(alg_id)
        rec[i, :, :] = astra.data2d.get(rec_id)
        astra.algorithm.delete(alg_id)
        astra.data3d.delete(sinogram_id)
    astra.data3d.delete(rec_id)
    return rec


if __name__ == '__main__':
    N_GPUS = 4
    with ThreadPool(N_GPUS) as p:
        result = p.starmap(fbp_subset, enumerate(np.split(proj_data, N_GPUS)))
        result = np.concatenate(result)
```

Note that you can just use a more lightweight 'threading' backend in joblib (not default) if all computations are on different GPUs.

**Comment by aljosahafner** [comment]
Excellent! Thanks a lot! :)  I think the issue can be closed!
